bn_mp_set_long_double.c

```diff
@@ -0,0 +1,276 @@
+     format although padded to a 96-bit format with 16 unused bits inserted between the
+     exponent and significand fields[15]). The follow-on Coldfire processors do not
+     support this 96-bit extended precision format.[16]
+
+     The x87 and Motorola 68881 80-bit formats meet the requirements of the IEEE 754
+     double extended format,[17] as does the IEEE 754 128-bit format."
+
+    There is also the use of implementation defined types in the bit field. That
+    restricts the range of compilers, too. Currently only GCC and CLANG (generic)
+    are allowed. It is unknown from which version on (tested with gcc 4.8 and clang
+    4.2.1 which are both quite old).
+
+    If there are any other problems: please use the third function here and contact
+    the authors.
+ */
+
+/*
+   Several architectures use "double" for "long double", we can use a shortcut in that case.
+*/
+#if ( (LDBL_MANT_DIG == DBL_MANT_DIG) && (LDBL_MAX_EXP == DBL_MAX_EXP) )
+int mp_set_long_double(mp_int *a, long double b)
+{
+   double _b = (double) b;
+   return mp_set_double(a, _b);
+}
+/*
+    The "long double" on a Sparc64 is either a quad-precision float with LDBL_MANT_DIG = 113
+    and LDBL_MAX_EXP = 16384, although in software only.
+*/
+#elif (    (LDBL_MANT_DIG == 64) && (LDBL_MAX_EXP == 16384) && (defined UINT64_MAX) \
+        && (defined LTM_NEARLY_IEC_559 ) \
+        && ( (defined __GNUC__) || (defined __clang__) ) \
+      )
+int mp_set_long_double(mp_int *a, long double b)
+{
+   uint64_t frac;
+   int exp, res = MP_OKAY;
+
+   /*
+      TODO:
+      This is not standard C, support of the types
+      uint64_t and uint16_t are implementation defined
+      if they do not translate to one of the allowed
+      types "int", "unsigned int", or "signed int"
+      (additionally "_Bool" in C11).
+    */
+   union {
+      long double ldbl;
+      /*
+         To cite a comment from libquadmath:
+         "On mingw targets the ms-bitfields option is active by default.
+          Therefore enforce gnu-bitfield style."
+         And who are we to ignore the advice of experts?
+      */
+      struct
+#ifdef __MINGW32__
+      __attribute__((gcc_struct))
+#endif
+      {
+         /* Set by GCC and Clang which are the only compilers allowed in here. */
+#if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__)
+         /* The m68k has some padding in the middle */
+#if ( (defined __m68k__) || (defined __MC68K__) || (defined M68000))
+         uint16_t sign:1;
+         uint16_t exponent:15;
+         uint16_t pad1:16;
+         uint64_t fraction:64;
+         uint16_t padrest;
+#endif
+#else /* __ORDER_LITTLE_ENDIAN__ */
+         uint64_t fraction:64;
+         uint16_t exponent:15;
+         uint16_t sign:1;
+         /* fill up (i368: to 12 bytes, x86_64: to 16 bytes */
+         uint16_t padding;
+#endif
+      } ldbl_guts;
+   } cast;
+
+   cast.ldbl = b;
+
+   exp = (int)cast.ldbl_guts.exponent;
```

Comment:
> Can we avoid using bitfields here?

Yes, we can (I think). But I found bit-fields simpler.

> bitfields are not well specified

They are quite well defined in the standard (vid. e.g.: 6.7.2.1 in INCITS/ISO/IEC 9899-2011[2012] for a start), but this one uses "or some other implementation-defined type[s]" so it is not "standard" in the most narrow sense, admitted but `long double` is not "fully standard" either.

That's why I have it restricted to the two compilers GCC and clang. The MingW port of GCC already needed some arm-twisting to make it complicit. MSVC can also be persuaded, at least that's what they told me, but I don't have one to test it and it is also moot because MSVC does not support `long double` as far as I remember.

And it is for `long double`, a type in widespread use but restricted to the x87 architecture. The number of different hardware implementations is: two. I think this code handles both of them sufficiently.

Yes, the world would be a happier place if `long double` goes extinct and gets replaced with a proper binary128. Or binary96 if you insist. But it won't happen, at least not in the near future. Maybe together with 128-bit registers?